File: src/main/java/ca/wglabs/streams/menu/MenuInquiry.java
package ca.wglabs.streams.menu;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Set;

public interface MenuInquiry {

    List<Dish> menu = Arrays.asList(new Dish("pork", false, 800, Dish.Type.MEAT),
                    new Dish("beef", false, 700, Dish.Type.MEAT),
                    new Dish("chicken", false, 400, Dish.Type.MEAT),
                    new Dish("french fries", true, 530, Dish.Type.OTHER),
                    new Dish("rice", true, 350, Dish.Type.OTHER),
                    new Dish("season fruit", true, 120, Dish.Type.OTHER),
                    new Dish("pizza", true, 550, Dish.Type.OTHER),
                    new Dish("prawns", false, 400, Dish.Type.FISH),
                    new Dish("salmon", false, 450, Dish.Type.FISH));


    List<String> getLowCaloricDishesNames();

    String getShortMenuCommaSeparated();

    Map<Dish.Type, List<Dish>> groupDishesByType();

    Map<Dish.CaloricLevel, List<Dish>> groupDishesByCaloricLevel();

    Map<Dish.Type, Map<Dish.CaloricLevel, List<Dish>>> groupDishedByTypeAndCaloricLevel();

    Map<Dish.Type, Long> countDishesInGroups();

    Map<Dish.Type, Long> sumCaloriesByType();

    Map<Dish.Type, Set<Dish.CaloricLevel>> caloricLevelsByType();

    Map<Boolean, List<Dish>> partitionByVegeterian();

    Map<Boolean, Map<Dish.Type, List<Dish>>> vegetarianDishesByType();

    Map mostCaloricPartitionedByVegetarian();
}

File: src/main/scala/ca/wglabs/collections/menu/MenuInquiry.scala
package ca.wglabs.collections.menu

import ca.wglabs.streams.menu.Dish.{CaloricLevel, Type}

trait MenuInquiry {

  val menu = List(
    Dish("pork", vegetarian = false, 800, Type.MEAT),
    Dish("beef", false, dtype = Type.MEAT, calories = 700),
    Dish("chicken", false, 400, Type.MEAT),
    Dish("french fries", true, 530, Type.OTHER),
    Dish("rice", true, 350, Type.OTHER),
    Dish("season fruit", true, 120, Type.OTHER),
    Dish("pizza", true, 550, Type.OTHER),
    Dish("prawns", false, 400, Type.FISH),
    Dish(vegetarian = false, calories = 450, dtype = Type.FISH, name = "salmon")
  )

  def getHighCaloricDishNames : List[String]

  def getShortMenuCommaSeparated : String

  def groupDishesByType : Map[Type, List[Dish]]

  def groupDishesByCaloricLevel : Map[CaloricLevel, List[Dish]]

  def groupDishesByTypeAndCaloricLevel : Map[Type, Map[CaloricLevel, List[Dish]]]

  def countDishesInGroups : Map[Type, Int]

  def sumCaloriesByType : Map[Type, Int]

  def caloricLevelsByType : Map[Type, Set[CaloricLevel]]

  def partitionByVegetarian : (List[Dish], List[Dish])

  def vegetarianDishesByType : (Map[Type, List[Dish]], Map[Type, List[Dish]])

  def mostCaloricPartitionedByVegetarian : (Dish, Dish)
}
